config-controller/pkg/client/client.go

```diff
@@ -0,0 +1,203 @@
+package client
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	storage "github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/roxctl/common"
+	"github.com/stackrox/rox/roxctl/common/auth"
+	roxctlIO "github.com/stackrox/rox/roxctl/common/io"
+	"github.com/stackrox/rox/roxctl/common/logger"
+	"github.com/stackrox/rox/roxctl/common/printer"
+)
+
+//go:generate mockgen-wrapper
+type CachedPolicyClient interface {
+	List(ctx context.Context) ([]*storage.Policy, error)
+	Get(ctx context.Context, name string) (*storage.Policy, bool, error)
+	Create(ctx context.Context, policy *storage.Policy) (*storage.Policy, error)
+	Update(ctx context.Context, policy *storage.Policy) error
+	FlushCache(ctx context.Context) error
+	EnsureFresh(ctx context.Context) error
+}
+
+//go:generate mockgen-wrapper
+type PolicyClient interface {
+	List(context.Context) ([]*storage.ListPolicy, error)
+	Get(ctx context.Context, id string) (*storage.Policy, error)
+	Post(context.Context, *storage.Policy) (*storage.Policy, error)
+	Put(context.Context, *storage.Policy) error
+}
+
+type grpcClient struct {
+	svc v1.PolicyServiceClient
+}
+
+func newGrpcClient(ctx context.Context) (PolicyClient, error) {
+	conn, err := common.GetGRPCConnection(auth.TokenAuth(), logger.NewLogger(roxctlIO.DefaultIO(), printer.DefaultColorPrinter()))
+	if err != nil {
+		return nil, errors.Wrap(err, "could not establish gRPC connection to Central")
+	}
+	svc := v1.NewPolicyServiceClient(conn)
+
+	return &grpcClient{
+		svc: svc,
+	}, nil
+}
+
+func (gc *grpcClient) List(ctx context.Context) ([]*storage.ListPolicy, error) {
+	allPolicies, err := gc.svc.ListPolicies(ctx, &v1.RawQuery{})
+	if err != nil {
+		return []*storage.ListPolicy{}, errors.Wrap(err, "Failed to list policies from grpc client")
+	}
+
+	return allPolicies.Policies, err
+}
+
+func (gc *grpcClient) Get(ctx context.Context, id string) (*storage.Policy, error) {
+	policy, err := gc.svc.GetPolicy(ctx, &v1.ResourceByID{Id: id})
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to fetch policy %s", id)
+	}
+
+	return policy, nil
+}
+
+func (gc *grpcClient) Post(ctx context.Context, policy *storage.Policy) (*storage.Policy, error) {
+	req := &v1.PostPolicyRequest{
+		Policy:                 policy,
+		EnableStrictValidation: true,
+	}
+
+	policy, err := gc.svc.PostPolicy(ctx, req)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to POST policy")
+	}
+
+	return policy, nil
+}
+
+func (gc *grpcClient) Put(ctx context.Context, policy *storage.Policy) error {
+	_, err := gc.svc.PutPolicy(ctx, policy)
+
+	if err != nil {
+		return errors.Wrap(err, "Failed to PUT policy")
+	}
+
+	return nil
+}
+
+type client struct {
+	svc         PolicyClient
+	cache       map[string]*storage.Policy
```

Comment:
Since reconciliation is done on a name basis (as opposed to ID), I chose to use it as a key.   This makes a small optimization by not having to iterate over the whole list of policies to find one that matches the name of the CR.

Reply:
Right, we are and have to be name based not id based here.